### PR TITLE
[Merged by Bors] - feat(Analysis/SpecialFunctions/Pow/NNReal): Specialise some lemmas to natural/integers powers

### PR DESCRIPTION
**Pull request by YaelDillies**
... and lemmas about negative real powers

From LeanAPAP

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean [hunk 1]
```diff
@@ -70,6 +84,43 @@ theorem rpow_add {x : ℝ≥0} (hx : x ≠ 0) (y z : ℝ) : x ^ (y + z) = x ^ y
...
+lemma rpow_sub_natCast' {n : ℕ} (h : y - n ≠ 0) (x : ℝ≥0) : x ^ (y - n) = x ^ y / x ^ n := by
+  ext; exact Real.rpow_sub_natCast' (mod_cast x.2) h
+
+lemma rpow_add_one (hx : x ≠ 0) (y : ℝ) : x ^ (y + 1) = x ^ y * x := by
+  simpa using rpow_add_natCast hx y 1
+
+lemma rpow_sub_one (hx : x ≠ 0) (y : ℝ) : x ^ (y - 1) = x ^ y / x := by
+  simpa using rpow_sub_natCast hx y 1
+
+-- TODO: Swap argument order to `rpow_add'`/`rpow_sub'`
```
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:117]: This looks like it's suggesting changing the argument order here to match `rpow_add'` and `rpow_sub'`. Is that what you mean, or do you want to change the argument order _of_ `rpow_add`' and `rpow_sub'`? 
If you do want to change the argument order here, why not just do it here? :)

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean [hunk 2]
```diff
@@ -224,6 +281,47 @@ theorem lt_rpow_inv_iff {x y : ℝ≥0} {z : ℝ} (hz : 0 < z) : x < y ^ z⁻¹
 theorem rpow_inv_lt_iff {x y : ℝ≥0} {z : ℝ} (hz : 0 < z) : x ^ z⁻¹ < y ↔ x < y ^ z := by
   simp only [← not_le, le_rpow_inv_iff hz]
 
+section
+variable {y : ℝ≥0}
+
+lemma rpow_lt_rpow_of_neg (hx : 0 < x) (hxy : x < y) (hz : z < 0) : y ^ z < x ^ z :=
```
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:287]: It looks like all 12 lemmas here aren't specialisations to natural or integer powers. Please add this to the PR description, or put these in a separate PR.

Comment (YaelDillies) [reply Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:287]: PR description updated

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean (hunk 1 above)
Comment (b-mehta) [reply Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean]: It seems you misunderstood my comment: The specific scenario I was excluding was exactly the change you made. Can you justify why you think the argument order should be changed here?

Comment (YaelDillies) [reply Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean]: The data `x : NNReal` was coming before the prop `y + z ≠ 0`, which is annoying for rewriting purposes as you always need to write one underscore before being able to provide the proof obligation

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean [hunk 3]
```diff
@@ -52,6 +52,8 @@ theorem rpow_eq_zero_iff {x : ℝ≥0} {y : ℝ} : x ^ y = 0 ↔ x = 0 ∧ y ≠
   rw [← NNReal.coe_inj, coe_rpow, ← NNReal.coe_eq_zero]
   exact Real.rpow_eq_zero_iff_of_nonneg x.2
 
+lemma rpow_eq_zero {x : ℝ≥0} (hy : y ≠ 0) : x ^ y = 0 ↔ x = 0 := by simp [hy]
```
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:55]:
```suggestion
lemma rpow_eq_zero (hy : y ≠ 0) : x ^ y = 0 ↔ x = 0 := by simp [hy]
```

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean [hunk 4]
```diff
@@ -60,16 +62,64 @@ theorem zero_rpow {x : ℝ} (h : x ≠ 0) : (0 : ℝ≥0) ^ x = 0 :=
...
+
 @[simp]
 theorem one_rpow (x : ℝ) : (1 : ℝ≥0) ^ x = 1 :=
   NNReal.eq <| Real.one_rpow _
 
 theorem rpow_add {x : ℝ≥0} (hx : x ≠ 0) (y z : ℝ) : x ^ (y + z) = x ^ y * x ^ z :=
   NNReal.eq <| Real.rpow_add (pos_iff_ne_zero.2 hx) _ _
 
-theorem rpow_add' (x : ℝ≥0) {y z : ℝ} (h : y + z ≠ 0) : x ^ (y + z) = x ^ y * x ^ z :=
+theorem rpow_add' {y z : ℝ} (h : y + z ≠ 0) (x : ℝ≥0) : x ^ (y + z) = x ^ y * x ^ z :=
```
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:84]:
```suggestion
theorem rpow_add' (h : y + z ≠ 0) (x : ℝ≥0) : x ^ (y + z) = x ^ y * x ^ z :=
```

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean [hunk 5]
```diff
@@ -81,17 +131,32 @@ lemma rpow_of_add_eq (x : ℝ≥0) (hw : w ≠ 0) (h : y + z = w) : x ^ w = x ^
...
+lemma rpow_mul_intCast (x : ℝ≥0) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y) ^ n := by
+  rw [rpow_mul, rpow_intCast]
 
 theorem rpow_neg_one (x : ℝ≥0) : x ^ (-1 : ℝ) = x⁻¹ := by simp [rpow_neg]
 
 theorem rpow_sub {x : ℝ≥0} (hx : x ≠ 0) (y z : ℝ) : x ^ (y - z) = x ^ y / x ^ z :=
   NNReal.eq <| Real.rpow_sub (pos_iff_ne_zero.2 hx) y z
 
-theorem rpow_sub' (x : ℝ≥0) {y z : ℝ} (h : y - z ≠ 0) : x ^ (y - z) = x ^ y / x ^ z :=
+theorem rpow_sub' {y z : ℝ} (h : y - z ≠ 0) (x : ℝ≥0) : x ^ (y - z) = x ^ y / x ^ z :=
```
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:151]:
```suggestion
theorem rpow_sub' (h : y - z ≠ 0) (x : ℝ≥0) : x ^ (y - z) = x ^ y / x ^ z :=
```

File: Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean [hunk 6]
```diff
@@ -412,13 +509,13 @@ theorem zero_rpow_mul_self (y : ℝ) : (0 : ℝ≥0∞) ^ y * (0 : ℝ≥0∞) ^
   exacts [zero_mul _, one_mul _, top_mul_top]
 
 @[norm_cast]
-theorem coe_rpow_of_ne_zero {x : ℝ≥0} (h : x ≠ 0) (y : ℝ) : (x : ℝ≥0∞) ^ y = (x ^ y : ℝ≥0) := by
+theorem coe_rpow_of_ne_zero {x : ℝ≥0} (h : x ≠ 0) (y : ℝ) : ↑(x ^ y) = (x : ℝ≥0∞) ^ y := by
```
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean:512]: Let's keep the explicit type ascription here for readability.

**Review by b-mehta — CHANGES_REQUESTED** [review]

**Review by b-mehta — COMMENTED** [review]
maintainer merge

**Comment by semorrison** [comment]
bors merge
